rust/lance-io/src/object_store/gcs_wrapper.rs

```diff
@@ -0,0 +1,458 @@
+        self.0.put(location, bytes).await
+    }
+
+    async fn put_opts(
+        &self,
+        location: &Path,
+        bytes: Bytes,
+        opts: PutOptions,
+    ) -> OSResult<PutResult> {
+        self.0.put_opts(location, bytes, opts).await
+    }
+
+    async fn put_multipart(
+        &self,
+        location: &Path,
+    ) -> OSResult<(MultipartId, Box<dyn AsyncWrite + Unpin + Send>)> {
+        // We don't return a real multipart id here. This will be addressed
+        // in object_store 0.10.0.
+        Upload::new(self.0.clone(), location.clone())
+            .map(|upload| (MultipartId::default(), Box::new(upload) as _))
+    }
+
+    async fn abort_multipart(&self, location: &Path, multipart_id: &MultipartId) -> OSResult<()> {
+        MultiPartStore::abort_multipart(self.0.as_ref(), location, multipart_id).await
+    }
+
+    async fn get_opts(&self, location: &Path, options: GetOptions) -> OSResult<GetResult> {
+        self.0.get_opts(location, options).await
+    }
+
+    async fn get_range(&self, location: &Path, range: Range<usize>) -> OSResult<Bytes> {
+        self.0.get_range(location, range).await
+    }
+
+    async fn get_ranges(&self, location: &Path, ranges: &[Range<usize>]) -> OSResult<Vec<Bytes>> {
+        self.0.get_ranges(location, ranges).await
+    }
+
+    async fn head(&self, location: &Path) -> OSResult<ObjectMeta> {
+        self.0.head(location).await
+    }
+
+    async fn delete(&self, location: &Path) -> OSResult<()> {
+        self.0.delete(location).await
+    }
+
+    fn delete_stream<'a>(
+        &'a self,
+        locations: BoxStream<'a, OSResult<Path>>,
+    ) -> BoxStream<'a, OSResult<Path>> {
+        self.0.delete_stream(locations)
+    }
+
+    fn list(&self, prefix: Option<&Path>) -> BoxStream<'_, OSResult<ObjectMeta>> {
+        self.0.list(prefix)
+    }
+
+    async fn list_with_delimiter(&self, prefix: Option<&Path>) -> OSResult<ListResult> {
+        self.0.list_with_delimiter(prefix).await
+    }
+
+    async fn copy(&self, from: &Path, to: &Path) -> OSResult<()> {
+        self.0.copy(from, to).await
+    }
+
+    async fn rename(&self, from: &Path, to: &Path) -> OSResult<()> {
+        self.0.rename(from, to).await
+    }
+
+    async fn copy_if_not_exists(&self, from: &Path, to: &Path) -> OSResult<()> {
+        self.0.copy_if_not_exists(from, to).await
+    }
+}
+
+enum UploadState {
+    Pending,
+    CreatingUpload(BoxFuture<'static, OSResult<MultipartId>>),
+    InProgress {
+        multipart_id: Arc<MultipartId>,
+        part_idx: u16,
+        futures: FuturesUnordered<
+            BoxFuture<'static, std::result::Result<(u16, PartId), UploadPutError>>,
+        >,
+        part_ids: Vec<Option<PartId>>,
+    },
+    PuttingSingle(BoxFuture<'static, OSResult<()>>),
+    Completing(BoxFuture<'static, OSResult<()>>),
+    Done,
+}
+
+/// Start at 5MB.
+const INITIAL_UPLOAD_SIZE: usize = 1024 * 1024 * 5;
```

Comment:
Any particular reason for this number?  Is this the minimum part size on GCS?

Reply:
Yeah this is generally the minimum part size.